### PR TITLE
Serialize models on the relation key instead of the modelName

### PR DESCRIPTION
**Pull request by jessedijkstra**
The models of a relations with serialized on the modelName. This makes it impossible to correctly serialize the relations.

See https://github.com/samselikoff/ember-cli-mirage/issues/905

**Todo:**
- [x] Fix and test the json-serializer
- [x] Fix and test the active-model-serializer

### CODE REVIEW

**Comment by jessedijkstra** [comment]
Saw that the serializer has been rewritten in it's entirety again. Rebasing now.

**Comment by jessedijkstra** [comment]
@samselikoff Rebased and fixed

**Comment by jessedijkstra** [comment]
@samselikoff Now fixed all the serialisation issues in the implemented serialisers. Would love to get some feedback on this PR with possible solutions that are less dependent on storing the key along with the resource when serialising associations.

This is quite a breaking bug and completely stops us from upgrading to the newest version of Mirage.

**Comment by samselikoff** [comment]
I think I'd rather not change all these method signatures, I know we spoke about making this code more clear overall but for this PR, could we just make the minimum change needed to get the tests passing?

Also if you want I can do this this weekend - your tests are valuable so let's keep them here.

**Comment by jessedijkstra** [comment]
@samselikoff Yeah, was looking for the minimum change. The biggest problem is that we loose the key the model is serialised on the moment we enter `getHashForIncludedResource`.

If you have multiple keys with the same model it's impossible to know what the key the resource should have based on the model definition.

Should we add the key to the resource rather then adding it as a `{ key, resource ]` pair? That does however require a change where we deepClone the model because it introduces resource depended state to the model making it unreliable and unsafe.

File: tests/integration/serializers/base/associations/sideloading-model-test.js [hunk 1]
```diff
@@ -91,7 +91,7 @@ test(`it can sideload a named has-many association`, function(assert) {
       name: 'Link',
       postIds: ['1']
     },
-    blogPosts: [
+    posts: [
```
Comment (samselikoff) [inline tests/integration/serializers/base/associations/sideloading-model-test.js]: I think this is wrong. Top-level collections should use the model name - what if this `blogPost` is also related to another model?

Attribute keys should only affect their model

**Comment by samselikoff** [comment]
I'm working on this now, I think I'll just push my changes up to your branch :)

File: tests/integration/serializers/base/associations/custom-key-collection-test.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  let blogPost = this.schema.blogPosts.find(1);
+  let result = registry.serialize(blogPost);
+
+  assert.deepEqual(result, {
+    blogPost: {
+      id: '1',
+      title: 'Lorem',
+      wordSmitheeId: '1'
+    },
+    wordSmithees: [
```
Comment (samselikoff) [inline tests/integration/serializers/base/associations/custom-key-collection-test.js]: This is also wrong

**Comment by samselikoff** [comment]
I think the bug is only for embedded, but you changed sideloaded as well.

Going to clean up the tests because I try to avoid adding new tests if we don't need them.

**Comment by samselikoff** [comment]
https://github.com/samselikoff/ember-cli-mirage/commit/1e4a5ef6cecf37cf6976328bb98705b4bee1d7d6 closes - thank you Jesse!

**Pull request closed without merging**